Implement initial Vector API expansion

gita-omr opened the pull request:
- create new optimization pass(VectorAPIExpansion)
- add VectorAPIExpansion to the cheap warm optimization strategy
- recognize and scalarize load(), store() and binaryOp() intrinsics
- add -Xjit command line option disableVectorAPIExpansion
- add -Xjit command line option traceVectorAPIExpansion

gita-omr commented:
Made WIP due to 2 sanity failures which are most likely not related. Investigating.

gita-omr commented:
Removed WIP since similar failures happened without this PR. Hoping somebody can take a look soon since there is a lot of code and we would like to get this into 0.28.

gita-omr commented:
@vijaysun-omr @0xdaryl @jdmpapin @r30shah @BradleyWood could you please review as much as you can?

gita-omr commented:
Noticed some GRA related issues. Fixing.

gita-omr commented:
Added commit with some fixes.

BradleyWood commented on runtime/compiler/env/j9method.cpp:
```diff
@@ -288,6 +288,12 @@ bool TR_ResolvedJ9Method::isMethodInValidLibrary()
    if (!strncmp(this->convertToMethod()->classNameChars(), "com/ibm/gpu/Kernel", 18))
       return true;
 
+   if (!strncmp(this->convertToMethod()->classNameChars(), "jdk/incubator/vector/FloatVector", 32))
```
strcmp() is probably better to avoid hard-coding the length

jdmpapin replied:
`strncmp()` is used because `classNameChars()` isn't necessarily NUL-terminated

jdmpapin replied:
@gita-omr, do you mean to accept the exact class name only, or any class name starting with this prefix (e.g. `FloatVector$1` in addition to `FloatVector` itself)? If you want only the exact matches, then you need to also compare `classNameLength()`

gita-omr replied:
Actually, it would be safe to accept all classes in jdk/incubator/vector/. Other checks around the code are similar. Maybe let's leave like that until we need to add more classes?

jdmpapin replied:
Personally I'd prefer if we could make the intention clear from the outset, because otherwise it will become hard to determine in the future. That's why I'm trying to comment on this as the case is added - it's certainly not obvious for many of the pre-existing cases here whether e.g. nested classes or other classes with a shared prefix might be matched, and whether it's important for them to match

Since all of jdk/incubator/vector/ could be accepted here, it's certainly not *necessary* to change it, but for clarity I would recommend one of the following:
- if you mean to accept just `FloatVector`, compare the length to limit it to exactly `FloatVector`; or
- if you mean to accept just `FloatVector` and its inner classes, compare the length and also separately accept anything that starts with `jdk/incubator/vector/FloatVector$`; or
- if you want to just start accepting everything in the vector package, change the prefix to `jdk/incubator/vector/`

If you really want to avoid doing any of those, could we have a comment in the code explaining that a prefix match is fine because any other classes that it accepts are also in the vector package?

gita-omr replied:
Yeah, I agree, it's easy to make it what it should be: `jdk/incubator/vector/`

gita-omr commented:
Changed to WIP to make sure Byte and Short types work properly.

gita-omr commented:
Added a commit to handle Byte and Short types. As @jdmpapin pointed out there were some issues with those. I decided to use the type promotion approach (see commit message).

gita-omr commented on runtime/compiler/optimizer/CMakeLists.txt:
```diff
@@ -74,6 +74,7 @@ j9jit_files(
 	optimizer/UnsafeFastPath.cpp
 	optimizer/VarHandleTransformer.cpp
 	optimizer/MethodHandleTransformer.cpp
+        optimizer/VectorAPIExpansion.cpp
```
Good find. Fixed.

r30shah commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -0,0 +1,1350 @@
+   
+   if (!disableVectorAPIExpansion &&
+       !TR::Compiler->om.canGenerateArraylets() &&
+       findVectorMethods())
```
Do we have guarded this opt to run only in java version from where methods from Vector JEPs are introduced. For example we do not need to run this optimization in Java 8 and Java 11. `findVectorMethods()` call would start walking tree tops to find if the method contains a call to the recognized vector API methods.

r30shah commented on runtime/compiler/optimizer/VectorAPIExpansion.hpp:
```diff
@@ -0,0 +1,816 @@
+/*******************************************************************************
+ * Copyright (c) 2000, 2021 IBM Corp. and others
```
Copyright year needs to be changed, Like wise in the previous `VectorAPIExpansion.cpp` file.

gita-omr replied:
I thought this is what we are supposed to specify...

BradleyWood replied:
I believe it should say "2021, 2021" with the first number indicating the original date of copywrite.

Also, the copywriter year in Optimizations.enum needs to be updated.

gita-omr replied:
Fixed!

gita-omr commented:
WIP: several fixes are coming and also would like to squash all the commits before merging.

gita-omr commented:
Thanks a lot for all the findings and suggestions!
I think I addressed all the issues except for the Java version. I could not find any examples in the code where we check it. Usually, features are based on some build macro or command line option. In the case of Vector API acceleration, I think it should be enabled by default, at least starting with Java 16. I was planning to use `ifVectorAPI` flag[1] that would be set during Inlining. This flag will be applied to the optimization itself as well as to the optimizations that will need to be added. Currently, the overhead is a quick pass through the treetops looking for the Vector API methods.

Please let me know what you think. @r30shah @jdmpapin @BradleyWood @0xdaryl @vijaysun-omr 

[1] https://github.com/eclipse/omr/pull/6117

vijaysun-omr commented:
Setting and checking IfVector sounds reasonable to me; a walk over the trees to collect this info may or may not even be needed, e.g. if the flag was set during IL gen itself (I believe this is how other flags are managed, e.g. IfNews) as the trees are being generated. 

Regardless, a walk over IL trees (if you figure that is what we need in the end) is also not really significant overhead even at warm.

fjeremic commented on runtime/compiler/optimizer/VectorAPIExpansion.cpp:
```diff
@@ -0,0 +1,1376 @@
+}
+
+
+// high level methods are disabled because they need excpetion handling to be inserted
+TR_VectorAPIExpansion::methodTableEntry
+TR_VectorAPIExpansion::methodTable[] =
+   {
+   {loadIntrinsicHandler,  TR::NoType, Vector,  {Unknown, elementType, numLanes}},                           // jdk_internal_vm_vector_VectorSupport_load
+   {storeIntrinsicHandler, TR::NoType, Unknown, {Unknown, elementType, numLanes, Unknown, Unknown, Vector}}, // jdk_internal_vm_vector_VectorSupport_store
+   {binaryIntrinsicHandler,TR::NoType, Vector,  {Unknown, Unknown, elementType, numLanes, Vector, Vector}},  // jdk_internal_vm_vector_VectorSupport_binaryOp
+   {unsupportedHandler /*fromArrayHandler*/,      TR::Float,  Vector,  {Species}}, // jdk_incubator_vector_FloatVector_fromArray,
+   {unsupportedHandler /*intoArrayHandler*/,      TR::Float,  Unknown, {Vector}},  // jdk_incubator_vector_FloatVector_intoArray,
+   {unsupportedHandler,    TR::Float,  Vector,  {Species}},                 // jdk_incubator_vector_FloatVector_fromArray_mask
+   {unsupportedHandler,    TR::Float,  Unknown, {}},                        // jdk_incubator_vector_FloatVector_intoArray_mask
+   {addHandler,            TR::Float,  Vector,  {Vector, Vector}},          // jdk_incubator_vector_FloatVector_add
+   {unsupportedHandler,    TR::Float,  Vector,  {}}                         // jdk_incubator_vector_VectorSpecies_indexInRange
+   };
```
This seems reliant on `Last_vector_api_method - First_vector_api_method` being a constant value at compile time. I think this code breaks if someone were to insert a new Vector API method without knowing they need to update this table.

We can fix this via `static_assert` ensuring that the number of elements here is exactly `Last_vector_api_method - First_vector_api_method`.

gita-omr replied:
Yeah, meant to do that and somehow forgot. Thanks for noticing!

gita-omr replied:
And will try the J2SE_V17 check. Thanks!

gita-omr replied:
Just pushed both

gita-omr commented:
Depends on https://github.com/eclipse/omr/pull/6136

fjeremic commented:
@gita-omr could you take a look at the linter + line endings job failures? They look related to this PR. Once we have that fixed we can launch sanity tests + any Vector API testing we have available via PR triggers.

gita-omr commented:
I believe Linter should pass now since eclipse/omr#6136 got propagated.

gita-omr commented:
Removed training spaces in all files from this PR. Please let me know when I can squash all the commits.

fjeremic commented:
```
12:34:15  + git diff --check origin/master HEAD
12:34:15  runtime/compiler/optimizer/VectorAPIExpansion.cpp:1381: new blank line at EOF.
```

Still one left. We should squash after that.

fjeremic commented:
Jenkins test sanity all jdk8,jdk17

gita-omr commented:
I am fixing AIX(XLC) build issue.

gita-omr commented:
Fixed xlC errors. Please let me know when to squash again.

fjeremic commented:
Jenkins test sanity all jdk8,jdk17

fjeremic commented:
Jenkins test sanity win jdk8

fjeremic commented:
Windows 32-bit failure is fixed via #13406.
